File: Kalman.py
class BatteryModel:
    def __init__(self, capacity, nominal_voltage):
        self.capacity = capacity  # Battery capacity in Ah
        self.voltage_nominal = nominal_voltage  # Nominal battery voltage in V

class KalmanFilter:
    def __init__(self, Q, R, initial_soc):
        self.Q = Q  # Process noise covariance
        self.R = R  # Measurement noise covariance
        self.x_hat = initial_soc  # Initial SOC estimate
        self.P = 0.01  # Initial estimation error covariance (adjust as needed)

    def predict(self, current):
        # Calculate change in SOC based on current
        delta_soc = -current / (battery.capacity * 3600)  # Convert current from mA to A and time to hours
        self.x_hat += delta_soc #update SOC estimation
        self.P += self.Q # Update estimation error covariance

    def update(self, voltage_measurement):
        K = self.P / (self.P + self.R) # Calculate Kalman gain
        self.x_hat += K * (voltage_measurement / battery.voltage_nominal - self.x_hat) # Update SOC estimate using voltage measurement
        self.P = (1 - K) * self.P # Update estimation error covariance

# Initialize battery model
battery = BatteryModel(capacity=30.0, nominal_voltage=59.2)

# Initialize Kalman filter with appropriate Q, R, and initial SOC
kalman_filter = KalmanFilter(Q=0.01, R=0.1, initial_soc=0.5)

# Simulate real-time input (replace this with actual data acquisition)
while(1):
    voltage_measurement = float(input("Enter voltage measurement (V): "))
    current_measurement = float(input("Enter current measurement (A): "))
    coulomb_counted_soc = kalman_filter.x_hat # Execute Coulomb counting to calculate SOC
    print("Coulomb Counting SOC: {:.2%}".format(coulomb_counted_soc))
    # Execute the Kalman filter
    kalman_filter.predict(current_measurement)
    kalman_filter.update(voltage_measurement)
    # Print the estimated SOC
    print("Kalman Filter SOC: {:.2%}".format(kalman_filter.x_hat))

File: kalman.c
#include <stdio.h>

// Battery model structure
typedef struct {
    double capacity;        // Battery capacity in Ah
    double voltage_nominal; // Nominal battery voltage in V
} BatteryModel;

// Kalman filter structure
typedef struct {
    double Q;     // Process noise covariance
    double R;     // Measurement noise covariance
    double x_hat; // Initial SOC estimate
    double P;     // Initial estimation error covariance (adjust as needed)
} KalmanFilter;

// Function to predict SOC based on current
void predict(KalmanFilter *kf, double current, BatteryModel *battery) {
    double delta_soc = -current / (battery->capacity * 3600); // Convert current from mA to A and time to hours
    kf->x_hat += delta_soc;  // Update SOC estimation
    kf->P += kf->Q;         // Update estimation error covariance
}

// Function to update SOC using voltage measurement
void update(KalmanFilter *kf, double voltage_measurement, BatteryModel *battery) {
    double K = kf->P / (kf->P + kf->R); // Calculate Kalman gain
    kf->x_hat += K * (voltage_measurement / battery->voltage_nominal - kf->x_hat); // Update SOC estimate using voltage measurement
    kf->P = (1 - K) * kf->P; // Update estimation error covariance
}

int main() {
    // Initialize battery model
    BatteryModel battery;
    battery.capacity = 30.0;
    battery.voltage_nominal = 59.2;

    // Initialize Kalman filter with appropriate Q, R, and initial SOC
    KalmanFilter kalman_filter;
    kalman_filter.Q = 0.01;
    kalman_filter.R = 0.1;
    kalman_filter.x_hat = 0.5;
    kalman_filter.P = 0.01;

    // Simulate real-time input (replace this with actual data acquisition)
    while (1) {
        double voltage_measurement;
        double current_measurement;
        double coulomb_counted_soc;

        // Get voltage and current measurements from user input
        printf("Enter voltage measurement (V): ");
        scanf("%lf", &voltage_measurement);
        printf("Enter current measurement (A): ");
        scanf("%lf", &current_measurement);

        // Execute Coulomb counting to calculate SOC
        coulomb_counted_soc = kalman_filter.x_hat;
        printf("Coulomb Counting SOC: %.2f%%\n", coulomb_counted_soc);

        // Execute the Kalman filter
        predict(&kalman_filter, current_measurement, &battery);
        update(&kalman_filter, voltage_measurement, &battery);

        // Print the estimated SOC
        printf("Kalman Filter SOC: %.2f%%\n", kalman_filter.x_hat);
    }

    return 0;
}
